### PR TITLE
Feat: Add mtls authen option to http

### PR DESCRIPTION
**Pull request by doubles-ss**
⚠️⚠️⚠️ Since we do not accept all types of pull requests and do not want to waste your time. Please be sure that you have read pull request rules:
https://github.com/louislam/uptime-kuma/blob/master/CONTRIBUTING.md#can-i-create-a-pull-request-for-uptime-kuma

Tick the checkbox if you understand [x]: 
- [x] I have read and understand the pull request rules.

# Description
Adds
- `tls_ca`, `tls_cert`, `tls_key` column in `monitor` table
- Add `mTLS` to the authentication method 

Fixes
- #139 

## Type of change

- User interface (UI)
- New feature (non-breaking change which adds functionality)

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I ran ESLint and other linters for modified files
- [x] I have performed a self-review of my own code and tested it
- [x] I have commented my code, particularly in hard-to-understand areas
  (including JSDoc for methods)
- [x] My changes generate no new warnings

## Screenshots (if any)
Screen Shot 2566-01-04 at 12 04 01
Screen Shot 2566-01-04 at 12 02 36

Please do not use any external image service. Instead, just paste in or drag and drop the image here, and it will be uploaded automatically.

### CODE REVIEW

**Comment by louislam** [comment]
Interesting, I wrongly thought the feature-request is about SSL/HTTPS. This is actually a auth method.

**Comment by doubles-ss** [comment]
@louislam May I ask you about your opinion about the certificate provided by the content of it or only provided files name?

**Comment by louislam** [comment]
> @louislam May I ask you about your opinion about the certificate provided by the content of it or only provided files name?

Content should be easier for docker users.

**Comment by doubles-ss** [comment]
> > @louislam May I ask you about your opinion about the certificate provided by the content of it or only provided files name?
> 
> Content should be easier for docker users.

OK, 

> > @louislam May I ask you about your opinion about the certificate provided by the content of it or only provided files name?
> 
> Content should be easier for docker users.

So, I think this way is easier for docker users to place content in textarea.

File: src/pages/EditMonitor.vue [hunk 1]
```diff
@@ -531,28 +531,47 @@
...
-                                    </div>
-
-                                    <div class="my-3">
-                                        <label for="basicauth" class="form-label">{{ $t("Password") }}</label>
-                                        <input id="basicauth-pass" v-model="monitor.basic_auth_pass" type="password" autocomplete="new-password" class="form-control" :placeholder="$t('Password')">
-                                    </div>
-                                    <template v-if="monitor.authMethod === 'ntlm' ">
+                                    <template v-if="monitor.authMethod === 'mtls' ">
+                                        <div class="my-3">
+                                            <label for="tls" class="form-label">{{ $t("Cert") }}</label>
```
Comment (AlexKraus) [inline src/pages/EditMonitor.vue:542]:
```suggestion
                                            <label for="tls-cert" class="form-label">{{ $t("Cert") }}</label>
```

File: src/pages/EditMonitor.vue [hunk 2]
```diff
@@ -531,28 +531,47 @@
...
+                                        <div class="my-3">
+                                            <label for="tls" class="form-label">{{ $t("CA") }}</label>
+                                            <textarea id="tls-ca" v-model="monitor.tlsCa" class="form-control" :placeholder="$t('Server CA')"></textarea>
+                                        </div>
+                                    </template>
+                                    <template v-else>
                                         <div class="my-3">
-                                            <label for="basicauth" class="form-label">{{ $t("Domain") }}</label>
-                                            <input id="basicauth-domain" v-model="monitor.authDomain" type="text" class="form-control" :placeholder="$t('Domain')">
+                                            <label for="basicauth" class="form-label">{{ $t("Username") }}</label>
```
Comment (AlexKraus) [inline src/pages/EditMonitor.vue:556]:
```suggestion
                                            <label for="basicauth-user" class="form-label">{{ $t("Username") }}</label>
```

File: src/pages/EditMonitor.vue [hunk 3]
```diff
@@ -531,28 +531,47 @@
...
+                                            <input id="basicauth-user" v-model="monitor.basic_auth_user" type="text" class="form-control" :placeholder="$t('Username')">
                                         </div>
 
                                         <div class="my-3">
-                                            <label for="basicauth" class="form-label">{{ $t("Workstation") }}</label>
-                                            <input id="basicauth-workstation" v-model="monitor.authWorkstation" type="text" class="form-control" :placeholder="$t('Workstation')">
+                                            <label for="basicauth" class="form-label">{{ $t("Password") }}</label>
```
Comment (AlexKraus) [inline src/pages/EditMonitor.vue:561]:
```suggestion
                                            <label for="basicauth-pass" class="form-label">{{ $t("Password") }}</label>
```

File: src/pages/EditMonitor.vue [hunk 4]
```diff
@@ -531,28 +531,47 @@
...
+                                            <input id="basicauth-pass" v-model="monitor.basic_auth_pass" type="password" autocomplete="new-password" class="form-control" :placeholder="$t('Password')">
                                         </div>
+                                        <template v-if="monitor.authMethod === 'ntlm' ">
+                                            <div class="my-3">
+                                                <label for="basicauth" class="form-label">{{ $t("Domain") }}</label>
```
Comment (AlexKraus) [inline src/pages/EditMonitor.vue:566]:
```suggestion
                                                <label for="ntlm-domain" class="form-label">{{ $t("Domain") }}</label>
```

File: src/pages/EditMonitor.vue [hunk 5]
```diff
@@ -531,28 +531,47 @@
...
+                                                <input id="basicauth-domain" v-model="monitor.authDomain" type="text" class="form-control" :placeholder="$t('Domain')">
```
Comment (AlexKraus) [inline src/pages/EditMonitor.vue:567]:
```suggestion
                                                <input id="ntlm-domain" v-model="monitor.authDomain" type="text" class="form-control" :placeholder="$t('Domain')">
```

File: src/pages/EditMonitor.vue [hunk 6]
```diff
@@ -531,28 +531,47 @@
...
+                                            </div>
+
+                                            <div class="my-3">
+                                                <label for="basicauth" class="form-label">{{ $t("Workstation") }}</label>
```
Comment (AlexKraus) [inline src/pages/EditMonitor.vue:571]:
```suggestion
                                                <label for="ntlm-workstation" class="form-label">{{ $t("Workstation") }}</label>
```

File: src/pages/EditMonitor.vue [hunk 7]
```diff
@@ -531,28 +531,47 @@
...
+                                                <input id="basicauth-workstation" v-model="monitor.authWorkstation" type="text" class="form-control" :placeholder="$t('Workstation')">
```
Comment (AlexKraus) [inline src/pages/EditMonitor.vue:572]:
```suggestion
                                                <input id="ntlm-workstation" v-model="monitor.authWorkstation" type="text" class="form-control" :placeholder="$t('Workstation')">
```

**Comment by louislam** [comment]
Maybe add `-----BEGIN CERTIFICATE-----` as the placeholder.

Any simple hello world app that supports mtls? So that I could test it quickly.

**Comment by AlexKraus** [comment]
Please check info@website.louislam.net regarding a temporary mtls example.

**Comment by louislam** [comment]
Cannot get it working, any ideas?

Error:

> 2023-03-12 18:57:25 - error:06065064:digital envelope routines:EVP_DecryptFinal_ex:bad decrypt

image

File: src/pages/EditMonitor.vue [hunk 8]
```diff
@@ -546,28 +546,47 @@
                                         <option value="ntlm">
                                             NTLM
                                         </option>
+                                        <option value="mtls">
+                                            mTLS
```
Comment (chakflying) [inline src/pages/EditMonitor.vue:550]:
```suggestion
                                            TLS Client Authentication
```

**Review by chakflying — COMMENTED** [review]
IMO this would be more clear.

**Comment by AlexKraus** [comment]
> Cannot get it working, any ideas?
> 
> Error:
> 
> > 2023-03-12 18:57:25 - error:06065064:digital envelope routines:EVP_DecryptFinal_ex:bad decrypt
> 
> image

Are you using the correct CA (letsencrypt ISRG Root X1 and R3 intermediary) (or none specific at all)?

**Comment by louislam** [comment]
I tested it without CA.

Now I also tested with these two certs (append both into the CA field), but it is also not working, same error.
https://letsencrypt.org/certs/isrg-root-x1-cross-signed.pem
https://letsencrypt.org/certs/lets-encrypt-r3.pem




Btw, I tested it with this docker image:

> docker run --rm -it -p 3000:3000 -p 3001:3001 --pull always -e 'UPTIME_KUMA_GH_REPO=doubles-ss:master' louislam/uptime-kuma:pr-test


I also tested on Windows, the error message is a little bit different:
> 2023-03-13 00:35:08 - error:1C800064:Provider routines::bad decrypt

**Comment by doubles-ss** [comment]
> I tested it without CA.
> 
> Now I also tested with these two certs (append both into the CA field), but it is also not working, same error. https://letsencrypt.org/certs/isrg-root-x1-cross-signed.pem https://letsencrypt.org/certs/lets-encrypt-r3.pem
> 
> 
> 
> Btw, I tested it with this docker image:
> 
> > docker run --rm -it -p 3000:3000 -p 3001:3001 --pull always -e 'UPTIME_KUMA_GH_REPO=doubles-ss:master' louislam/uptime-kuma:pr-test
> 
> I also tested on Windows, the error message is a little bit different:
> 
> > 2023-03-13 00:35:08 - error:1C800064:Provider routines::bad decrypt

Look like your private key has encrypted and need to use passphase to decrypt your key.
Please remove your passphase first via `openssl pkcs8 -topk8 -inform pem -in  encrypted.key -out private_pkcs8.pem -outform pem -nocrypt`

**Comment by louislam** [comment]
Thanks, it is working now.

**Comment by AlexKraus** [comment]
You are right. My bad! I thought there was a field for key passphrase and prepared the example accordingly.

**Pull request merged**